architectures/decentralized/testing/src/docker_setup.rs

```diff
@@ -167,6 +169,106 @@ pub async fn spawn_new_client(docker_client: Arc<Docker>) -> Result<String, Dock
+/// This allows rejoining with the same identity after disconnecting
+pub async fn spawn_client_with_keypair(
+    docker_client: Arc<Docker>,
+    host_keypair_path: &str,
+) -> Result<String, DockerWatcherError> {
+    // Set the container name based on the ones that are already running.
+    let new_container_name = get_name_of_new_client_container(docker_client.clone()).await;
+
+    // Check if GPU is available
+    let has_gpu = has_gpu_support();
+
+    // Setting extra hosts and optionally nvidia request
+    let network_name = "test_psyche-test-network";
+
+    // Create bind mount for the keypair
+    let container_keypair_path = "/root/.config/solana/id.json";
+    let binds = vec![format!("{}:{}", host_keypair_path, container_keypair_path)];
+
+    let host_config = if has_gpu {
+        // Setting nvidia usage parameters
+        let device_request = DeviceRequest {
+            driver: Some("nvidia".to_string()),
+            count: Some(1),
+            capabilities: Some(vec![vec!["gpu".to_string()]]),
+            ..Default::default()
+        };
+
+        HostConfig {
+            device_requests: Some(vec![device_request]),
+            extra_hosts: Some(vec!["host.docker.internal:host-gateway".to_string()]),
+            network_mode: Some(network_name.to_string()),
+            binds: Some(binds),
+            ..Default::default()
+        }
+    } else {
+        HostConfig {
+            extra_hosts: Some(vec!["host.docker.internal:host-gateway".to_string()]),
+            network_mode: Some(network_name.to_string()),
+            binds: Some(binds),
+            ..Default::default()
+        }
+    };
+
+    // Get env vars from config file
+    let env_vars: Vec<String> = std::fs::read_to_string("../../../config/client/.env.local")
+        .expect("Failed to read env file")
+        .lines()
+        .filter_map(|line| {
+            let line = line.trim();
+            if line.is_empty() || line.starts_with("#") {
+                None
+            } else {
+                Some(line.to_string())
+            }
+        })
+        .collect();
+    let envs = if has_gpu {
+        [env_vars, vec!["NVIDIA_DRIVER_CAPABILITIES=all".to_string()]].concat()
+    } else {
+        env_vars
+    };
+
+    // Override entrypoint to skip solana-keygen (which would overwrite the mounted keypair)
+    // We run the same commands as client_test_entrypoint.sh but WITHOUT solana-keygen new
+    let entrypoint_cmd = "set -o errexit && \
+         solana config set --url \"${RPC}\" && \
+         solana airdrop 10 \"$(solana-keygen pubkey)\" && \
+         psyche-solana-client train \
+             --wallet-private-key-path \"/root/.config/solana/id.json\" \
+             --rpc \"${RPC}\" \
+             --ws-rpc \"${WS_RPC}\" \
+             --run-id \"${RUN_ID}\" \
+             --logs \"json\""
+        .to_string();
+    let entrypoint = vec!["/bin/sh", "-c", &entrypoint_cmd];
```

Comment:
maybe we should just move this to a separate bash script and pass it as entrypoint